src/influence_diagram.jl

```diff
@@ -811,151 +807,134 @@ julia> add_utilities!(diagram, "V1", [0, -25])
+diagram.D and diagram.V and storing appropriate values to them.
 
 # Examples
 ```julia
 generate_arcs!(diagram)
 ```
 """
 function generate_arcs!(diagram::InfluenceDiagram)
-
-    # Chance and decision nodes
-    C_and_D = filter(x -> !isa(x, ValueNode), diagram.Nodes)
+    C_and_D = filter(x -> !isa(x[2], ValueNode), pairs(diagram.Nodes)) # Collects all nodes not ValueNodes
     n_CD = length(C_and_D)
-    # Value nodes
-    V_nodes = filter(x -> isa(x, ValueNode), diagram.Nodes)
-    n_V = length(V_nodes)
-
-    validate_structure(diagram.Nodes, C_and_D, n_CD, V_nodes, n_V)
-
-    # Declare vectors for results (final resting place InfluenceDiagram.Names and InfluenceDiagram.I_j)
-    Names = Vector{Name}(undef, n_CD+n_V)
-    I_j = Vector{Vector{Node}}(undef, n_CD+n_V)
-    states = Vector{Vector{Name}}()
-    S = Vector{State}(undef, n_CD)
-    C = Vector{Node}()
-    D = Vector{Node}()
-    V = Vector{Node}()
-
-    # Declare helper collections
-    indices = Dict{Name, Node}()
-    indexed_nodes = Set{Name}()
-    # Declare index
-    index = 1
-
-
-    while true
-        # Index nodes C and D that don't yet have indices but whose I_j have indices
-        new_nodes = filter(j -> (j.name ∉ indexed_nodes && Set(j.I_j) ⊆ indexed_nodes), C_and_D)
-        for j in new_nodes
-            # Update helper collections
-            push!(indices, j.name => index)
-            push!(indexed_nodes, j.name)
-            # Update results
-            Names[index] = Name(j.name)    #TODO datatype conversion happens here, should we use push! ?
-            I_j[index] = map(x -> Node(indices[x]), j.I_j)
-            push!(states, j.states)
-            S[index] = State(length(j.states))
-            if isa(j, ChanceNode)
-                push!(C, Node(index))
-            else
-                push!(D, Node(index))
-            end
-            # Increase index
-            index += 1
-        end
-
-        # If no new nodes were indexed this iteration, terminate while loop
-        if isempty(new_nodes)
-            if index < n_CD
-                throw(DomainError("The influence diagram should be acyclic."))
-            else
-                break
-            end
+    V_ = filter(x -> isa(x[2], ValueNode), pairs(diagram.Nodes)) # Collects all ValueNodes
+    n_V = length(V_)
+
+    validate_structure(diagram.Nodes, C_and_D, n_CD, V_, n_V)
+
+    # Declare vectors for results
+    I_j = OrderedDict{Name, Vector{Name}}()
+    states = OrderedDict{Name, Vector{Name}}()
+    S = OrderedDict{Name, State}()
+    C = OrderedDict{Name, ChanceNode}()
+    D = OrderedDict{Name, DecisionNode}()
+    V = OrderedDict{Name, ValueNode}()
+
+    #LINE BELOW WAS NEEDED FOR REORDERING OF NODES, UNLIKE WHAT WE THOUGHT
+    diagram.Nodes = merge(C_and_D, V_)
+
+    #Assigning indices for all nodes (by constructing new nodes, because node-structs are immutable)
+    node_index = 1
+    for (name, node) in diagram.Nodes
```

Comment:
Not that we'd really need it yet, but is diagram.Nodes guaranteed to be in topological order? It would probably be convenient if the indices represented a topological ordering of the diagram